src/cargo/ops/vendor.rs

```diff
@@ -0,0 +1,338 @@
+        let ws = Workspace::new(&extra, ws.config())?;
+        extra_workspaces.push(ws);
+    }
+    let workspaces = extra_workspaces.iter().chain(Some(ws)).collect::<Vec<_>>();
+    let vendor_config =
+        sync(ws.config(), &workspaces, opts).chain_err(|| format!("failed to sync"))?;
+
+    let shell = ws.config().shell();
+    if shell.verbosity() != Verbosity::Quiet {
+        eprint!("To use vendored sources, add this to your .cargo/config for this project:\n\n");
+        print!("{}", &toml::to_string(&vendor_config).unwrap());
+    }
+
+    Ok(())
+}
+
+#[derive(Serialize)]
+struct VendorConfig {
+    source: BTreeMap<String, VendorSource>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "lowercase", untagged)]
+enum VendorSource {
+    Directory {
+        directory: PathBuf,
+    },
+    Registry {
+        registry: Option<String>,
+        #[serde(rename = "replace-with")]
+        replace_with: String,
+    },
+    Git {
+        git: String,
+        branch: Option<String>,
+        tag: Option<String>,
+        rev: Option<String>,
+        #[serde(rename = "replace-with")]
+        replace_with: String,
+    },
+}
+
+fn sync(
+    config: &Config,
+    workspaces: &[&Workspace<'_>],
+    opts: &VendorOptions<'_>,
+) -> CargoResult<VendorConfig> {
+    let canonical_destination = opts.destination.canonicalize();
+    let canonical_destination = canonical_destination
+        .as_ref()
+        .map(|p| &**p)
+        .unwrap_or(opts.destination);
+
+    fs::create_dir_all(&canonical_destination)?;
+    let mut to_remove = HashSet::new();
+    if !opts.no_delete {
+        for entry in canonical_destination.read_dir()? {
+            let entry = entry?;
+            to_remove.insert(entry.path());
+        }
+    }
+
+    // First up attempt to work around rust-lang/cargo#5956. Apparently build
+    // artifacts sprout up in Cargo's global cache for whatever reason, although
+    // it's unsure what tool is causing these issues at this time. For now we
+    // apply a heavy-hammer approach which is to delete Cargo's unpacked version
+    // of each crate to start off with. After we do this we'll re-resolve and
+    // redownload again, which should trigger Cargo to re-extract all the
+    // crates.
+    //
+    // Note that errors are largely ignored here as this is a best-effort
+    // attempt. If anything fails here we basically just move on to the next
+    // crate to work with.
+    for ws in workspaces {
+        let (packages, resolve) =
+            ops::resolve_ws(&ws).chain_err(|| "failed to load pkg lockfile")?;
+
+        packages
+            .get_many(resolve.iter())
+            .chain_err(|| "failed to download packages")?;
+
+        for pkg in resolve.iter() {
+            // Don't delete actual source code!
+            if pkg.source_id().is_path() {
+                if let Ok(path) = pkg.source_id().url().to_file_path() {
+                    to_remove.remove(&path);
```

Comment:
The bug with Windows is that `to_remove` contains `\\?` prefixed paths, whereas the source_id returns a normal path.